src/main/java/com/ably/kafka/connect/utils/RecordHeaderConversions.java

```diff
@@ -180,5 +180,16 @@ MessageExtras toMessageExtras(){
             return null;
         }
 
+        /**
+         * Stringify header value if it is primitive.
+         */
+        private static Object stringifyHeaderValueIfPrimitive(Header header) {
+            if (header.schema() != null && header.schema().type().isPrimitive()) {
```

Comment:
Only handling primitives presumably still leaves the problem for headers that kafka auto-types as nonprimitives. So I guess arrays, maps, structs, and who knows what else, containing arbitrary and arbitrarily nested data. (Which will then be rejected by realtime, which validates that headers is flat).

Is the issue that nonprimitives might not have a sensible String.valueOf()? If so could we json-serialize nonprimitives or something?

Reply:
That's a good question, I thought we only have problems with primitives. `String.valueOf()` probably will produce something sensible, but it won't be JSON, if that's OK I can do this. 

We also have additional logic for push header we intentionally check for `Map` and for `String` (looks like Kafka can produce both for JSON value) and if it is `Map` we  json-serialize it. Then we parse result string as `JsonObject` and put it in extras. 

Should we try to do this for any header value? Or is it very specific exception?

Reply:
That's specific to push, since there'd otherwise be no way of populating the extras.push struct. No need to do that for anything other than push. extras.push is tightly structured and validated for push notification metadata, and that needs to work, so extras.push is the one exception that can/should end up as a deep object, if the customer uses the magic "com.ably.extras.push" header key. So that shouldn't be changed here.

(...I'm not sure I see the point of serializing the push header Map to json and then immediately deserializing it again, but whatever.)

If String.valueOf() will do something sensible I'd be inclined to just do that for everything (other than push).